[ONNX] Speed up export of large models

ilyasher opened the pull request:
This commit speeds up the ONNX export of large models by making the following changes:

- Remove unecessary memcpy in `GetGraphProtoSize`
- In `export.cpp`, pass around a pointer to the ModelProto instead of the ModelProto itself.

The shape inference function is still the slowest part of the export for these models with large weights taking up 50% or more of the export time.

ilyasher committed:
[ONNX] Speed up export of large models

This commit speeds up the ONNX export of large models
by making the following changes:

- Remove unecessary memcpy in GetGraphProtoSize
- In export.cpp, pass around a pointer to the ModelProto
  instead of the ModelProto itself.
- The shape inference function is very time-consuming,
  so only call it once when exporting (instead of twice).

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -686,7 +687,7 @@ def _optimize_graph(
     graph = _C._jit_pass_canonicalize(graph)
     _C._jit_pass_lint(graph)
-    if GLOBALS.onnx_shape_inference:
+    if GLOBALS.onnx_shape_inference and not skip_shape_inference:
```
I'm a little unsure about this. I recall the motivation is to correct shape/type after passes like `onnx_peephole`.

ilyasher replied:
OK, I've removed this change.

BowenBao reviewed (commented):
We should profile shape_type_inference and see why it's so slow. Maybe somewhere we are doing too many copies or too many unnecessary constant propagations.

ilyasher commented:
> We should profile shape_type_inference and see why it's so slow. Maybe somewhere we are doing too many copies or too many unnecessary constant propagations.

Here's what I found while I was profiling shape inference:
1. During shape inference, each node is converted into its own separate ONNX graph to run ONNX's shape inference. Creating these ONNX nodes is expensive because it requires copying the weights.
2. The constant folding step always makes a copy of the output tensor.

As an example, here is what happens to a `torch.nn.Linear` layer when it is exported:
1. The layer is turned into Transpose -> Matmul ONNX ops. (fast)
2. The weights are copied into a new ONNX Transpose node in order to perform ONNX shape inference (slow)
3. Constant folding is run on the Transpose node and the transposed tensor is cloned and stored in ConstantValueMap (slow)
4. Now we're done with the Transpose, and move on to the Matmul.

I don't see an easy way around these copies.

thiagocrepaldi reviewed (commented):
Publicly disabling by default the shape inference seems very tricky.

At most `GLOBALS.onnx_shape_inference` could be distilled in more fine grained options that users could change on their own runs, but keeping the default as strict as possible

ilyasher committed:
Don't skip 2nd shape inference in export

Signed-off-by: Ilya Sherstyuk <isherstyuk@nvidia.com>

ilyasher commented:
> Publicly disabling by default the shape inference seems very tricky.

Shape inference wasn't entirely disabled, but I changed it to only run once at the end instead of once in the middle and once in the end.

In any case, I reverted that change in response to a different comment.

BowenBao commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -833,32 +833,26 @@ unsigned long long int GraphEncoder::GetGraphProtoSize(
+
+    // Don't actually copy the buffer into tensor_proto since that is expensive.
+    // All we actually need is its size.
+    size += tensor_proto->ByteSizeLong();
```
This is fine as is. I'm thinking since this is rather an estimate of GraphProto size, it feels fine to drop `tensor_proto` and only compute size from tensor? And then further drop GraphProto copy on L836?

One can give this a try in a follow-up as well as fix the bug that this function currently neglects constants in the model which could be large. Related #102479, #100873

ilyasher replied:
I guess it's unlikely, but maybe these non-weight fields could also push the proto to be over 2GB.

ilyasher commented:
@pytorchbot merge

thiagocrepaldi commented:
@pytorchbot merge
